src/mumble/pluginAPI/PluginComponents.h

```diff
@@ -0,0 +1,133 @@
+/**
+ * This header file contains definitions of types and other components used in Mumble's plugin system
+ */
+
+#ifndef MUMBLE_PLUGINCOMPONENT_H_
+#define MUMBLE_PLUGINCOMPONENT_H_
+
+#include <stdint.h>
+
+#define STATUS_OK 0
+
+struct Version {
+	int32_t major;
+	int32_t minor;
+	int32_t patch;
+};
+
+
+struct MumbleUser {
+	const char *name;
+	uint32_t id;
+};
+
+struct MumbleChannel {
+	const char *name;
+	int32_t id;
+};
+
+
+typedef struct Version Version_t;
+typedef struct MumbleUser MumbleUser_t;
+typedef struct MumbleChannel MumbleChannel_t;
+
+typedef int32_t MumbleConnection_t
+
+
+struct MumbleAPI {
+	// -------- Memory (de-)allocation --------
+	
+	/// Allocates memory for a MumbleUser. Once the user struct is no longer needed, it has to be freed by calling
+	/// freeMumbleUser.
+	///
+	/// @returns A pointer to the allocated struct.
+	MumbleUser_t* (*allocateMumbleUser)();
```

Comment:
I don't think it's a good idea to provide a allocator function for every single type which might be passed through.
Also why does the API user want to allocate a user?

Reply:
> I don't think it's a good idea to provide a allocator function for every single type which might be passed through.

I was following @NouberNou's suggestion on how memory magement could be done. Currently the memory management is completely up to the plugin and the need to allocate the used structs first, might ensure that the Dev actually thinks about freeing them later.

Tbh I'm not quite happy with the current state either, but I guess the only other way would be to create the structs on the heap and return pointers to those. Then the plugin would have to free the memory afterwards. However as memory should be freed by the same process that allocated it, it had to pass it to Mumble again. Since we're not having objects that contain a constructor, we'd have to have a free-method for every struct that is being returned (at least as far as I see it - feel free to correct me).

> Also why does the API user want to allocate a user?

Currently the intended workflow is to allocate an empty user, pass that into an API function that will fill the user struct's fields with meaningful data that can then be used by the plugin. Once finished with it, the plugin can free the user struct by calling the respective function

Reply:
In reference to my comments further below, why do we need structs at all. All structs (the two) I've seen so far are pretty much useless.

> that will fill the user struct's fields with meaningful data
What is meaningful data? Currently it always retrieves the users name, even if the API user doesn't even need/want the name, are you just gonna add all details you can think of about the user into there? That would be a huge waste of processing time to collect all the info that the plugin doesn't need.
I'd prefer to keep it like teamspeak api, just make a getName function to get the name.
That could then be somewhat reused for server/channel name too.
Or maybe a generic "getUserVariable" function, like teamspeak does. With an enum specifying what value you want.
Then we have getUser/Channel/ServerVariable functions, which a enum as parameter.
Maybe want to specialize these for return type (string, float, int)

Reply:
> why do we need structs at all

Well that was probably me trying hard to use OO programming (or rather something remotely feeling like it) where it can't be done properly and where it might be smarter not to :thinking:

Reply:
I rewrote the API to not rely on structs - If you have the time I'd appreciate another review on it :)